---
title: "Huanan market samples"
author: "Jacques van Helden"
date: "`r Sys.Date()`"
output:
  html_document:
    self_contained: no
    code_download: false
    code_folding: "hide"
    fig_caption: yes
    highlight: zenburn
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
    number_sections: no
  html_notebook:
  html_fragment:
    code_download: false
    fig_width: 7
    fig_height: 6
    fig_caption: true
    toc: yes
    number_sections: false
    smart: true
    keep_md: true
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
subtitle: "Analysis of the results from Bloom (2023)"
editor_options:
  chunk_output_type: console
---

```{r settings, include=FALSE, echo=FALSE, eval=TRUE}
#### Libraries ####
message("Loading required libraries")

requiredLib <- c(
  "knitr",
  "RColorBrewer"
)

for (lib in requiredLib) {
  if (!require(lib, character.only = TRUE)) {
    install.packages(lib, )
  }
  require(lib, character.only = TRUE)
}


## Knitr options
options(width = 300)
# options(encoding = 'UTF-8')
knitr::opts_chunk$set(
  fig.width = 7, fig.height = 5, 
  fig.path = 'figures/huanan_',
  fig.align = "center", 
  size = "tiny", 
  echo = FALSE, 
  eval = TRUE, 
  warning = FALSE, 
  message = FALSE, 
  results = TRUE, 
#  knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file()), 
  comment = "")

# setwd(rprojroot::find_rstudio_root_file())

options(scipen = 12) ## Max number of digits for non-scientific notation
# knitr::asis_output("\\footnotesize")

## Take a copy of the original parameters to restore them after each figure
par_ori <- par(no.readonly = TRUE)

```

## Data loading

This report presents an analysis of the results published by Jesse D. Bloom (2023), based on the data from the Chinese Center for Disease Control (Liu et al., 2023). 


The data was imported from Jesse Bloom's github repository. 

- <https://github.com/jbloom/Huanan_market_samples>

```{r input_dirs-and_files}
#### Parameters ####

parameters <- list()

## Directories
parameters$dirs <- list()

parameters$dirs$base <- ".." ## Paths are relative to the src directory

## Result dir
parameters$dirs$results <- file.path(parameters$dirs$base, "results")
if (is.numeric(parameters$selected_week)) {
  parameters$dirs$results <- file.path(parameters$dirs$results, date_suffix)
}
dir.create(parameters$dirs$results, showWarnings = FALSE, recursive = TRUE)




## Input files
infiles <- list()
infiles$mito_composition_per_sample <- file.path(parameters$dirs$base, "results", "aggregated_counts", "mito_composition_by_sample.csv")

## Epsilon for log10 conversion
parameters$epsilon <- 0.1

```



```{r load_data}
## Load reads per sample from Bloom, 2023

mito_composition_per_sample <- read.csv(infiles$mito_composition_per_sample, header = TRUE, sep = ",", quote = c())
# dim(mito_composition_per_sample)
# View(mito_composition_per_sample)
# names(mito_composition_per_sample)


## Compute additional columns for convenience
mito_composition_per_sample$species_common_name <- paste0(mito_composition_per_sample$species, " (", 
                                                          mito_composition_per_sample$common_name, ")")
mito_composition_per_sample$aligned_reads_log10 <- log10(mito_composition_per_sample$aligned_reads + 
                                                           parameters$epsilon)
mito_composition_per_sample$SARS2_aligned_reads_log10 <- log10(mito_composition_per_sample$SARS2_aligned_reads + parameters$epsilon)

stat <- list()
stat$n_samples <- length(unique(mito_composition_per_sample$sample))
stat$n_species <- length(unique(mito_composition_per_sample$species))

# n_common_names <- length(unique(mito_composition_per_sample$common_name))

```

Number of samples: `r stat$n_samples`

Number of species: `r stat$n_species`


## Number of samples per species

```{r samples_per_species, fig.width=12, fig.height=6, out.width="100%", fig.cap="Number of positive samples per species (top) and of different species per sample (bottom)"}


samples_per_species <- sort(table(subset(mito_composition_per_sample, aligned_reads > 0, select = "species_common_name")), decreasing = TRUE)

par(mar=c(12,5,4,1))
barplot(samples_per_species, 
        main = "Number of positive sample per species",
        ylab = "Number of positive samples", 
        col = "#BBBBFF",
        ylim=c(0,180),
        cex.names = 0.7, las = 2)
abline(h = stat$n_samples)

par(par_ori, no.readonly = TRUE)
```

## Number of species per sample

```{r species_per_sample, fig.width=12, fig.height=6, out.width="100%", fig.cap="Number of positive samples per species (top) and of different species per sample (bottom)"}

par(mar=c(5,5,4,1))
barplot(sort(table(subset(mito_composition_per_sample, aligned_reads > 0, select = "sample")), decreasing = TRUE), 
        main = "Number of species per sample",
        ylab = "Number of species", 
        col = "#BBFFFF",
        ylim=c(0,40),
        cex.names = 0.3, las = 2)
abline(h = stat$n_species)


par(par_ori, no.readonly = TRUE)
```



## Samples vs species (common name)

```{r counts_per_samples_vs_species_heatmap, fig.width=12, fig.height=12, out.width="100%", fig.cap="Heatmap showing the number of aligned reads per sample (rows) and species common name (columns). Counts were log10-converted to highlight the full range of values. Zero counts are displayed in yellow. "}
#### Counts per sample and per species

## Compute a contingency table with the number of reads per samples and common_name ##
aligned_per_sample_vs_species_common_name <- xtabs(aligned_reads ~ sample + species_common_name, data  = mito_composition_per_sample)
# colnames(sars2_per_sample_vs_species_common_name)

## Add a column with the number of SARS-CoV-2 matching reads per sample
## Note that the SARS-CoV-2 matching reads are listed in a separate column, in a redundant way, We thus just keep the first column
sars2_per_sample_vs_species_common_name <- xtabs(SARS2_aligned_reads ~ sample + species_common_name, data  = mito_composition_per_sample)
# range(apply(sars2_per_sample_vs_species_common_name, 1, var))
sars2_per_sample <- sars2_per_sample_vs_species_common_name[,1]

# class(aligned_per_sample_vs_species_common_name)
# dim(aligned_per_sample_vs_species_common_name)
aligned_per_sample_vs_species_common_name <- as.data.frame.matrix(aligned_per_sample_vs_species_common_name)
# class(aligned_per_sample_vs_species_common_name)
# dim(aligned_per_sample_vs_species_common_name)
aligned_per_sample_vs_species_common_name[, "SARS-CoV-2"] <- sars2_per_sample[rownames(aligned_per_sample_vs_species_common_name)]
# View(aligned_per_sample_vs_species_common_name)
# class(aligned_per_sample_vs_species_common_name)
# dim(aligned_per_sample_vs_species_common_name)
# aligned_per_sample_vs_species_common_name$`SARS-CoV-2`

# colnames(aligned_per_sample_vs_species_common_name)

## Compute log10 of the counts per sample/common_name
aligned_per_sample_vs_species_common_name_log10 <- log10(aligned_per_sample_vs_species_common_name + parameters$epsilon)
# dim(aligned_per_sample_vs_species_common_name_log10)

## Cluster species according to reads per sample
species_clust <- hclust(dist(t(aligned_per_sample_vs_species_common_name_log10)), method = "complete")
# plot(species_clust)

## Cluster samples according to reads per species
sample_clust <- hclust(dist(aligned_per_sample_vs_species_common_name_log10), method = "complete")
# plot(sample_clust)

colfunc <- colorRampPalette(c("white", "blue"))
blue.scale <- colfunc(n = 128)
gray.scale <- gray.colors(n = 128, start = 0, end = 1, rev=TRUE)

## Draw the heatmap ##
heatmap(as.matrix(aligned_per_sample_vs_species_common_name_log10), 
        main = "Aligned reads (log10 counts) per sample and species",
        scale = "none", 
        margins = c(16,8), 
        xlab = "Species common name", 
        ylab  = "Sample", 
        Rowv = as.dendrogram(sample_clust), 
        Colv = as.dendrogram(species_clust),
        cexCol = 0.8, cexRow = 0.3,
        col = c("#FFFFDD", blue.scale))



```




```{r aligned_counts_per_sample}
#### Aligned counts per sample
aligned_per_sample <- apply(aligned_per_sample_vs_species_common_name, 1, sum)
aligned_per_species <- apply(aligned_per_sample_vs_species_common_name, 2, sum)

```


## Correlations between mitochondrial genome and SARS-CoV-2 read counts

```{r mito_vs_sars2, fig.width=8, fig.height=12, out.width="100%", fig.cap="Number of reads (log10-transformed) aligned on mitochondrial genome of selected species (abcsissa) versus SARS-CoV-2 (ordinate)."}

species_of_interest <- c("human", "cow", "raccoon dog", "snakehead fish", "Largemouth bass", "carp")
selected_species <- "human"

par(mfrow=c(3,2))

for (selected_species in species_of_interest) {

  ## Select the subset of samples for the species of interest
  species_subset <- subset(mito_composition_per_sample, common_name == selected_species)
  species_common_name <- unique(species_subset$species_common_name)

  ## Select the subset having strickly positive counts for both the species of interest and SARS-CoV-2
  nozero <- (species_subset$aligned_reads > 0) & (species_subset$SARS2_aligned_reads > 0)
  nozero_subset <- species_subset[nozero, ]
  # View(nozero_subset)
  # length(species_subset$aligned_reads*species_subset$SARS2_aligned_reads > 0)
  # dim(species_subset)
  # nrow(nozero_subset)
  
  species_subset$color <- "#000088"
  species_subset[species_subset$aligned_reads == 0, "color"] <- "#880000"
  species_subset[species_subset$SARS2_aligned_reads == 0, "color"] <- "#008800"
    
  ## Compute correlation between species counts and SARS-CoV-2 counts
  species_cor <- cor(species_subset$aligned_reads, species_subset$SARS2_aligned_reads)
  nozero_species_cor <- cor(nozero_subset$aligned_reads, nozero_subset$SARS2_aligned_reads)
  species_log10_cor <- cor(species_subset$aligned_reads_log10, species_subset$SARS2_aligned_reads_log10)
  nozero_species_log10_cor <- cor(nozero_subset$aligned_reads_log10, nozero_subset$SARS2_aligned_reads_log10)
  
  ## Scatter plot
  plot(species_subset$aligned_reads_log10,
       species_subset$SARS2_aligned_reads_log10,
       main = species_common_name,
       col = species_subset$color,
       xlab = "Aligned on mitochondrial genome (log10 counts)",
       ylab = "Aligned on SARS-CoV-2, log10(counts)",
       las=1)
  # points(nozero_subset$aligned_reads_log10,
  #        nozero_subset$SARS2_aligned_reads_log10,
  #        pch=2, col="black")
  abline(h=-1:7, col="#DDDDDD")
  abline(v=-1:4, col="#DDDDDD")
  abline(h = 0)
  abline(v = 0)
  legend("topright", cex = 0.9,
         legend = c(
           paste0("cor=", round(species_cor, 3)),
           paste0("nonzero cor=", round(nozero_species_cor, 3)),
           paste0("log10 cor=", round(species_log10_cor, 3)),
           paste0("nozero log10 cor=", round(nozero_species_log10_cor, 3))
         ))
}

par(par_ori, no.readonly = TRUE)
```

## References

Bloom JD. Association between SARS-CoV-2 and metagenomic content of samples from the Huanan Seafood Market. Microbiology, 2023 DOI: [10.1101/2023.04.25.538336](https://doi.org/10.1101/2023.04.25.538336).

Liu WJ, Liu P, Lei W, et al. Surveillance of SARS-CoV-2 at the Huanan Seafood Market. Nature 2023; published online April 5. DOI: [10.1038/s41586-023-06043-2](https://doi.org/10.1038/s41586-023-06043-2).

